### PR TITLE
Climber

### PR DESCRIPTION
**Pull request by JoeJanickiChSh**
Implement current limits into the Climber subsystem which measures the motors' current and uses that value to stop the motors instead of using limit switches. Three new commands are added: extendCurrent, retractCurrent, and moveCurrent, which implement this new functionality.

### CODE REVIEW

**Comment by msoucy** [comment]
Blocking on chopshop-166/chopshoplib#98

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 1]
```diff
@@ -17,6 +18,7 @@
   // Constants:
   private final double EXTEND_SPEED = 1.0;
   private final double RETRACT_SPEED = -1.0;
+  private final double CURRENT_LIMIT = 1.0; // Current limit in amps, motor stops when this is reached
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Climber.java]: Real current draw should be closer to 30A so we probably want to make this a bit higher than that.
image

This should also come from the robot map as it depends on the specific robot design.

**Review by bot190 — CHANGES_REQUESTED** [review]
I think we probably want to combine these sets of commands so that if either the limit switches are triggered OR the motor draws too much current we will stop moving.

File: src/main/java/frc/robot/subsystems/Climber.java (hunk 1 above)
Comment (JoeJanickiChSh) [reply src/main/java/frc/robot/subsystems/Climber.java]: These changes should be fixed once motor validators get merged into chopshoplib.

Comment (msoucy) [reply src/main/java/frc/robot/subsystems/Climber.java]: Now available if you update chopshoplib's version in `build.gradle` to `2022.5.1`

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 2]
```diff
@@ -16,6 +16,25 @@
...
 
 public class GastonMap extends RobotMap {
+    // Climber
+    private final DigitalInput leftUpperLimit = new DigitalInput(0);
+    private final DigitalInput leftLowerLimit = new DigitalInput(1);
+    private final DigitalInput rightUpperLimit = new DigitalInput(2);
+    private final DigitalInput rightLowerLimit = new DigitalInput(3);
+
+    private final PIDSparkMax leftMotor = new PIDSparkMax(0, MotorType.kBrushless);
+    private final PIDSparkMax rightMotor = new PIDSparkMax(1, MotorType.kBrushless);
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/GastonMap.java]: These should be declared within the `getLeftTelescopeMap` and its right equivalent.

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 3]
```diff
@@ -122,10 +115,14 @@ public IntakeMap getIntakeMap() {
     }
 
     public TelescopeMap getLeftTelescopeMap() {
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/GastonMap.java]: Are they "climber" maps, or "telescope" maps?

File: src/main/java/frc/robot/maps/RobotMap.java [hunk 4]
```diff
@@ -145,10 +145,11 @@ public TelescopeMap(SmartMotorController motor, BooleanSupplier upperLimit, Bool
             this.motor = motor;
             this.upperLimit = upperLimit;
             this.lowerLimit = lowerLimit;
+
         }
 
         public TelescopeMap() {
-            this(new SmartMotorController(), new MockDigitalInput(), new MockDigitalInput());
+            this(new SmartMotorController(new SmartMotorController()), new MockDigitalInput(), new MockDigitalInput());
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/RobotMap.java]: 🚩

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 5]
```diff
@@ -19,50 +18,104 @@
...
     }).onEnd((interrupted) -> {
       motor.set(0.0);
     });
   }
 
   public CommandBase extend() {
-    return startEnd("Extend", () -> {
-      motor.set(upperModifier.applyAsDouble(EXTEND_SPEED));
-    }, () -> {
+    return cmd("Extend").onExecute(() -> {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Climber.java:66]: Can these three and their equivalent `retract` commands be abstracted at all? They look very similar, except for the speed.

**Review by msoucy — CHANGES_REQUESTED** [review]

**Review by msoucy — DISMISSED** [review]

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 6]
```diff
@@ -73,30 +68,18 @@ public SwerveDriveMap getSwerveDriveMap() {
...
     public IntakeMap getIntakeMap() {
         // PID coefficients
         // initializes relative encoder and pid controller, we don't need the encoder rn
 
         // private RelativeEncoder deploymentEncoder =
         // deploymentMotor.getEncoder().getRaw();
+        final DigitalInput outsideLimit = new DigitalInput(4);
+        final DigitalInput insideLimit = new DigitalInput(5);
+
+        final PIDSparkMax deploymentMotor = new PIDSparkMax(2, MotorType.kBrushless);
```
Comment (jeffaburke) [inline src/main/java/frc/robot/maps/GastonMap.java]: **_PLEASE_** Change your motor IDs

**Review by jeffaburke — CHANGES_REQUESTED** [review]

**Review by bot190 — APPROVED** [review]

**Review by msoucy — APPROVED** [review]

**Review by jeffaburke — APPROVED** [review]

**Pull request merged**